Clarify that arrays and registers of size 0 are allowed

garrison opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.

Are you changing the specification? This PR needs to be approved by the TSC members.

Are you changing the grammar to adjust to the specification?

The specification is the source of truth and any grammar updates should also coincide with accepted specification updates in the same PR.

-->

### Summary

Fixes #532.

### Details and comments

The primary change is to make it so a "size" can be a non-negative integer instead of a positive integer.  This works for qubits as well as classical bits/registers, but it leads to some clarifications being necessary for some other types:

- Integers of zero bits are explicitly forbidden by this PR.  An ~~unsigned~~ integer must have at least one bit.  ~~A signed integer must have at least two bits (the sign bit plus one integer bit).  The latter was probably implicitly assumed before, but now it is explicit.~~  NOTE: I saw nothing specifying that an integer must have a "normal" number of bits, i.e., 8, 16, 32, or 64, even though the examples tend to use these values.
- Floating point numbers must have a number of bits that corresponds to a valid IEEE 754 floating point type, so I believe it is unnecessary to explicitly say that zero bits are forbidden.
- Like unsigned integers, angles must have at least one bit.
- I added a statement permitting arrays to have dimension(s) which are of length zero.  I avoided the word "length" because it is currently not used to describe the individual dimensions given by the shape.  I am open to rephrasing this with guidance.
 
#### Remaining action items

- [x] add a release note
- [x] add tests for the parser
- [x] get tests to pass
- [x] revise release note: make it more specific

jlapeyre commented on source/language/types.rst:
```diff
@@ -218,6 +218,7 @@ There are n-bit signed and unsigned integers. The statements ``int[size] name;``
 signed 1:n-1:0 and unsigned 0:n:0 integers of the given size. The sizes
 and the surrounding brackets can be omitted (*e.g.* ``int name;``) to use
 a precision that is specified by the particular target architecture.
+If provided, the ``size`` of an unsigned integer must be at least 1, and the ``size`` of an unsigned integer must be at least 2.
```
```suggestion
If provided, the ``size`` of an unsigned integer must be at least 1, and the ``size`` of an signed integer must be at least 2.
```

jwoehr replied:
**"a"** signed

jlapeyre commented:
> I saw nothing specifying that an integer must have a "normal" number of bits 8, 16, 32, or 64

Compare to Julia and Rust which are defined by an implementation rather than a spec (although Rust probably follows a more coherent design) . Julia `Int13` and Rust `i13` are not forbidden, but neither are the implemented and no one is interested in implementing them. The story in C is of course more complicated. But some motivations are shared with OQ3 since programs must be able to run on platforms with limited resources with varying widths of native types.

In OQ3, apparently a fully compliant implementation must handle 13 bit integers. I assume restricting the width of data types in OQ3 has been discussed in the past. Prima facie, it makes sense to restrict the widths somehow. For example require a minimum collection of data types, and allow implementations to support others.

jlapeyre commented:
> A signed integer must have at least two bits (the sign bit plus one integer bit). The latter was probably implicitly assumed before, but now it is explicit.

Otherwise, we would have a third representation of boolean values.

jakelishman commented:
Fwiw, a signed integer need not have two bits; `int[1]` can represent 0 and -1 in two's complement and +0 and -0 in one's complement. Not useful, but still well defined.

garrison commented:
> Fwiw, a signed integer need not have two bits; `int[1]` can represent 0 and -1 in two's complement and +0 and -0 in one's complement. Not useful, but still well defined.

I updated to allow for this in 19cd9fd3244f229962a37f75ae9e7f3287b19d97, as the spec is simpler this way.  We can always continue to discuss during the next TSC meeting.

levbishop commented:
This was discussed at TSC meeting and broadly approved. We'd like to have a test for the parser to confirm compliance with the given spec changes.

blakejohnson commented:
@garrison we are still hoping to see a test for the parser here.

garrison commented:
The one interesting thing I have found so far is that if a user attempts to provide an init expression for an empty array, it currently results in a `QASM3ParsingError`.  Of course, there is nothing to actually initialize in these cases, so there is no particular reason for the user to provide an init expression, but it still seems unexpected to me.  Some examples:

```openqasm
array[int, 0] a = {};
```
```openqasm
array[int, 2, 0] a = {{}, {}};
```
```openqasm
array[int, 0, 2] a = {};
```

On the other hand, the following program parses successfully, even though the dimension does not match:

```openqasm
array[int, 0] a = {0};
```

jakelishman commented:
I _think_ its unambiguous with the rest of the grammar - none of the other rules that involve `{}` are valid in the only locations that the array literal is (iirc).

garrison commented:
I modified the parser so that the tests that I added would pass in 8713166f1930a4.  I did this in the way most obvious to me.  I am very much a beginner with ANTLR, so this is likely shoddy work, but I wanted nonetheless to make it public for review while I try to figure out how to make it better.

levbishop reviewed (approved):
Thanks for seeing this through. DIscussed at 2024-12-11 TSC meeeting
